### PR TITLE
Autograd updated

### PR DESCRIPTION
**Pull request by annzhukovets**

### CODE REVIEW

File: interface.py [hunk 1]
```diff
@@ -10,160 +13,261 @@ class Numeric(Interface):
...
-        else:
-            return 0
+    def derivative(self, variable):
+        if hasattr(self.arg, 'grad') and isinstance(form_of(self.subject), Operator):
+            self.arg.grad = variable.sum()
+            form_of(self.subject).derivative(resolve(variable))
+        elif hasattr(self.subject, 'grad') and form_of(self.arg).derivative(resolve(variable)):
+            self.subject.grad = variable.sum()
+            form_of(self.arg).derivative(resolve(variable))
+        elif hasattr(self.arg, 'grad') and hasattr(self.subject, 'grad'):
```
Comment (haydnv) [inline interface.py:97]: should this case go first?

File: interface.py [hunk 2]
```diff
@@ -10,160 +13,261 @@ class Numeric(Interface):
...
+            self.arg.grad = variable.sum()
```
Comment (haydnv) [inline interface.py:98]: It seems strange to me that the gradient isn't calculated until the `derivative` method is called, because it assumes that the calling code has already evaluated the `derivative` of every preceding `Operator`. Since the value of `grad` depends on `variable`, maybe the gradient should be passed as an argument to `derivative`?

File: interface.py [hunk 3]
```diff
@@ -10,160 +13,261 @@ class Numeric(Interface):
...
+class MatMul(Dual):
+    def __resolve__(self):
+        return resolve(self.subject) @ resolve(self.arg)
+
+    def derivative(self, variable):
+        if hasattr(self.arg, 'grad') and isinstance(form_of(self.subject), Operator):
+            self.arg.grad = resolve(self.subject).T @ variable
+            form_of(self.subject).derivative(variable @ (resolve(self.arg).T))
+        elif hasattr(self.subject, 'grad') and isinstance(form_of(self.arg), Operator):
+            self.subject.grad = variable @ (resolve(self.arg).T)
```
Comment (haydnv) [inline interface.py:123]: Or maybe there should be a container class called `OperatorGraph` that computes the gradient for each `Variable` in the graph?

File: interface.py [hunk 4]
```diff
@@ -10,160 +13,261 @@ class Numeric(Interface):
...
-            return arg
-        else:
-            return 0
+    def derivative(self, variable):
+        if hasattr(self.arg, 'grad') and isinstance(form_of(self.subject), Operator):
+            self.arg.grad = variable.sum()
+            form_of(self.subject).derivative(resolve(variable))
+        elif hasattr(self.subject, 'grad') and form_of(self.arg).derivative(resolve(variable)):
+            self.subject.grad = variable.sum()
+            form_of(self.arg).derivative(resolve(variable))
```
Comment (haydnv) [inline interface.py:96]: Note that calling `resolve` outside of `__resolve__` means that the return value of `derivative` won't be valid except for the one time it's called--would it be better to let the calling code call `resolve(x.derivative())`? That way code like `derivative(derivative(x))` would still be valid.

File: ml.py [hunk 5]
```diff
@@ -25,40 +153,70 @@ def train(self, inputs, labels):
...
+        self.lr = learning_rate
+
+    def train(self, inputs, labels, cost):
+        outputs = self.model.eval(inputs)
+        assert isinstance(form_of(outputs), Operator)
+        loss = cost(resolve(outputs), resolve(labels))
+        dloss = cost(resolve(outputs), resolve(labels), dloss=True)
+        form_of(outputs).derivative(dloss)
+
+        for layer in self.model.layers:
```
Comment (haydnv) [inline ml.py:187]: it would be better if this didn't assume a `Sequential` model, but it's fine to leave that change for a future PR

File: ml.py [hunk 6]
```diff
@@ -25,40 +153,70 @@ def train(self, inputs, labels):
...
+            if layer.weights:
+                delta = layer.weights.grad
```
Comment (haydnv) [inline ml.py:189]: I think it would be better to just calculate this `grad` value in the `backprop` function, that way `Operator.derivative` won't have side-effects (which it shouldn't because the derivative of a function is always the same)

Comment (haydnv) [reply ml.py:189]: Alternately, you could make `Variable.grad` a function which depends on some state of the `Operator` calling it

File: test_optimizer_2.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+from pickletools import optimize
```
Comment (haydnv) [inline test_optimizer_2.py:1]: I think you didn't mean to include this import

File: test_optimizer_2.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+batch = 8
+threshold = 0.001
+inputs = np.random.rand(batch, 15)
+label =  np.expand_dims((inputs*2).sum(axis=1), -1)
+model = Sequential([(15, 8, None), (8, 20, None), (20, 1, None)])
+
+optimizer = GradientDescent2(model, learning_rate = 1e-4)
+
+def mse(predict, target, dloss=False):
```
Comment (haydnv) [inline test_optimizer_2.py:15]: Instead of the `dloss=True` option, why not just call `derivative(mse(outputs, labels))`?

**Review by haydnv — CHANGES_REQUESTED** [review]

File: ml.py (hunk 5 above)
Comment (haydnv) [reply ml.py:143]: I'll merge this change in and add a TODO

File: ml.py (hunk 6 above)
Comment (haydnv) [reply ml.py:145]: I'll also add a TODO for this

**Review by haydnv — APPROVED** [review]
Please fix the conflicts with the main branch and then merge this in
